checks/license_check.go

```diff
@@ -0,0 +1,146 @@
+const LicenseCheckPolicy = "LicenseCheck"
+
+//nolint:gochecknoinits
+func init() {
+	registerCheck(LicenseCheckPolicy, LicenseCheck)
+}
+
+const (
+	copying      = "copy(ing|right)"
+	license      = "(un)?licen[sc]e"
+	preferredExt = "*\\.(md|markdown|html)$"
+	anyExt       = ".[^./]"
+	ofl          = "ofl"
+	patents      = "patents"
+)
+
+// regex converted from https://github.com/licensee/licensee/blob/master/lib/licensee/project_files/license_file.rb
+var (
+	extensions  = []string{"xml", "go", "gemspec"}
+	regexChecks = []checks{
+		{rstr: copying, f: nil},
+		{rstr: license, f: nil},
+		{rstr: license + preferredExt, f: nil},
+		{rstr: copying + preferredExt, f: nil},
+		{rstr: copying + anyExt, f: nil},
+		{rstr: ofl, f: nil},
+		{rstr: ofl + preferredExt, f: nil},
+		{rstr: patents, f: nil},
+		{rstr: license, f: extensionMatch, p: []string{"spdx", "header"}},
+		{rstr: license + "[-_][^.]*", f: extensionMatch, p: extensions},
+		{rstr: copying + "[-_][^.]*", f: extensionMatch, p: extensions},
+		{rstr: "\\w+[-_]" + license + "[^.]*", f: extensionMatch, p: extensions},
+		{rstr: "\\w+[-_]" + copying + "[^.]*", f: extensionMatch, p: extensions},
+		{rstr: ofl, f: extensionMatch, p: extensions},
+	}
+)
+
+// extensionMatch to check for matching extension.
+func extensionMatch(f string, exts []string) bool {
+	s := strings.Split(f, ".")
+
+	if len(s) <= 1 {
+		return false
+	}
+
+	fext := s[len(s)-1]
+
+	found := false
+	for _, ext := range exts {
+		if ext == fext {
+			found = true
+			break
+		}
+	}
+
+	return found
+}
+
+// testLicenseCheck used for testing purposes.
+func testLicenseCheck(name string) bool {
+	return checkLicense(name)
+}
+
+// LicenseCheck runs LicenseCheck check.
+func LicenseCheck(c *checker.CheckRequest) checker.CheckResult {
+	var r bool
+
+	onFile := func(name string, dl checker.DetailLogger, data FileCbData) (bool, error) {
+		pdata := FileGetCbDataAsBoolPointer(data)
+
+		if checkLicense(name) {
+			c.Dlogger.Info3(&checker.LogMessage{
+				Path:   name,
+				Type:   checker.FileTypeSource,
+				Offset: 1,
+			})
+			*pdata = true
+			return false, nil
+		}
+		return true, nil
+	}
+
+	err := CheckIfFileExists(LicenseCheckPolicy, c, onFile, &r)
+	if err != nil {
+		return checker.CreateRuntimeErrorResult(LicenseCheckPolicy, err)
+	}
+	if r {
+		return checker.CreateMaxScoreResult(LicenseCheckPolicy, "license file detected")
+	}
+	return checker.CreateMinScoreResult(LicenseCheckPolicy, "license file not detected")
+}
+
+// checkLicense to check whether the name parameter fulfill license file criteria.
+func checkLicense(name string) bool {
+	for _, check := range regexChecks {
+		rg := regexp.MustCompile(check.rstr)
+
+		nameLower := strings.ToLower(name)
+		t := rg.MatchString(nameLower)
+		if t {
+			extFound := false
+
+			// check extension calling f function
+			if check.f != nil {
+				extFound = check.f(nameLower, check.p)
+			} else {
+				// got a match
+				extFound = true
```

Comment:
I think the assumption here is that `check.f` is always `extensionMatch`, correct? If that's the case, please add a comment.

Reply:
Updated the PR with more comments.